### PR TITLE
feat(prepare-transactions): add estimated gas fee

### PR DESCRIPTION
**Pull request by jeanregisser**
### Description

This updates `prepareTransactions` to be able to return the estimated gas fee, closer to the real fee.

For swaps, it uses the `estimatedGasUse` returned by our API if present.

See ["Optimizing network fee for swaps"](https://www.notion.so/valora-inc/Optimizing-network-fee-for-swaps-561e70c8d5b34eb7bb74f93e70d02227) for more details.

Next up: UI changes for swaps making use of the new estimated gas fee.

### Test plan

- Updated tests

### Related issues

- Fixes RET-991

### Backwards compatibility

Yes

### CODE REVIEW

File: src/viem/estimateFeesPerGas.ts [hunk 1]
```diff
@@ -17,14 +17,27 @@ export async function estimateFeesPerGas(
       getMaxPriorityFeePerGas(client, feeCurrency),
     ])
     const maxFeePerGas = gasPrice + maxPriorityFeePerGas
-    return { maxFeePerGas, maxPriorityFeePerGas }
+    // TODO: properly calculate baseFeePerGas
+    return { maxFeePerGas, maxPriorityFeePerGas, baseFeePerGas: BigInt(maxFeePerGas) }
```
Comment (jeanregisser) [inline src/viem/estimateFeesPerGas.ts:21]: I'll address this in a follow up PR. We need to call the GasPriceMinimum contract to get the right value based on the fee currency set.

File: src/viem/prepareTransactions.ts [hunk 2]
```diff
@@ -14,29 +14,37 @@ import {
...
 const TAG = 'viem/prepareTransactions'
 
 // Supported transaction types
-export type TransactionRequest = TransactionRequestCIP42 | TransactionRequestEIP1559
+export type TransactionRequest = (TransactionRequestCIP42 | TransactionRequestEIP1559) & {
+  // Custom fields needed for showing the user the estimated gas fee
+  // underscored to denote that they are not part of the TransactionRequest fields from viem
+  // and only intended for internal use in Valora
+  _estimatedGasUse?: bigint
+  _baseFeePerGas?: bigint
```
Comment (jeanregisser) [inline src/viem/prepareTransactions.ts:32]: Let me know what you think of this.

File: src/viem/prepareTransactions.ts [hunk 3]
```diff
@@ -61,6 +69,24 @@ export function getMaxGasFee(txs: TransactionRequest[]): BigNumber {
...
+    }
+    if (!tx._baseFeePerGas || !tx.maxFeePerGas) {
+      throw new Error('Missing _baseFeePerGas or maxFeePerGas')
+    }
+    const expectedFeePerGas = tx._baseFeePerGas + (tx.maxPriorityFeePerGas ?? BigInt(0))
+    estimatedGasFee +=
+      estimatedGas * (expectedFeePerGas < tx.maxFeePerGas ? expectedFeePerGas : tx.maxFeePerGas)
+  }
+  return new BigNumber(estimatedGasFee.toString())
+}
```
Comment (jeanregisser) [inline src/viem/prepareTransactions.ts:88]: The formula implemented is this:

`estimatedGasFee = (estimatedGasUse ?? gas) * min(baseFeePerGas + (maxPriorityFeePerGas ?? 0), maxFeePerGas)`

File: src/viem/prepareTransactions.ts (hunk 2 above)
Comment (kathaypacific) [reply src/viem/prepareTransactions.ts:32]: seems fine to me

**Review by kathaypacific — APPROVED** [review]
LGTM!! 🚀

File: src/viem/estimateFeesPerGas.ts (hunk 1 above)
Comment (jeanregisser) [reply src/viem/estimateFeesPerGas.ts:21]: Addressed in https://github.com/valora-inc/wallet/pull/4796

**Pull request merged**